privilege-request-protocol.md

```diff
@@ -0,0 +1,208 @@
+HTTP Privilege Request Protocol
+===============================
+
+Introduction
+------------
+This memo specifies a protocol between an HTTP client and server for the
+client to request additional access privilege when denied access to a resource
+by the server for insufficient privilege.
+
+The client will typically use some form of authentication when communicating
+with the server. The method and semantics of authentication are beyond the
+scope of this memo. Methods could include, but are not limited to, use of the
+HTTP `Authorization` header using a scheme supported by the server such as
+`Bearer`, or a TLS client certificate.
+
+A client may comprise a software application component that is identified in
+some way to the server. The method by which a software application is identified
+to the server is beyond the scope of this memo. Methods could include, but
+are not limited to, use of the HTTP `Origin` header, or an application identity
+associated in some way with the `Bearer` token.
+
+The method by which a server determines whether an HTTP client has sufficient
+privilege to access the requested resource is beyond the scope of this memo.
+Methods could include, but are not limited to, determining whether the
+authenticated identity is a member of a configured role group, or whether the
+user associated with the authenticated identity has endorsed the identified
+software application currently being used for the access attempted.
+
+The method by which a server determines whether it's appropriate to allow the
+client to request additional privilege is beyond the scope of this memo.
+Methods could include, but are not limited to, determining whether the
+authenticated identity is the controller of the resource, or is the owner of
+the server.
+
+The key words "**MUST**", "**MUST NOT**", "**REQUIRED**", "**SHALL**", "**SHALL
+NOT**", "**SHOULD**", "**SHOULD NOT**", "**RECOMMENDED**", "**NOT RECOMMENDED**",
+"**MAY**", and "**OPTIONAL**" in this document are to be interpreted as
+described in BCP 14 \[[RFC2119][]\] \[[RFC8174][]\] when, and only when, they
+appear in all capitals, as shown here.
+
+Operation
+---------
+The following is a typical exchange according to this protocol:
+
+1. The client attempts access to a resource.
+2. The server determines the client has insufficient privilege and denies the
+   request, returning an HTTP `403 Forbidden` response. The server further
+   determines it's appropriate to allow the client to request additional
+   privilege, so includes in the response a `Link` for the privilege request
+   endpoint URI.
+3. The client determines to request additional privilege and makes a simple
+   `POST` to the provided privilege request endpoint URI without credentials.
```

Comment:
there should be CORS headers on this endpoint, so that the POST can be done in an XHR request from a web app, right?

Reply:
CORS is orthogonal to this protocol. presumably the same amount of CORS the server is already doing to talk to the client would be done here. the example interaction includes appropriate CORS headers. i will add at least a mention of CORS in the "syntax" section.